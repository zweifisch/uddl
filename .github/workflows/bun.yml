name: Bun

on:
  push:
    branches: [ "main" ]
  pull_request:
    branches: [ "main" ]

jobs:
  build:

    runs-on: ubuntu-latest

    steps:
    - uses: actions/checkout@v4
    - name: Setup Bun
      uses: oven-sh/setup-bun@v2.0.1
    - name: Install Dependencies
      run: bun install --frozen-lockfile
    - name: Run tests
      run: bun test
