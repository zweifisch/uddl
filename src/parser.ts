import { createToken, Lexer, CstParser  } from "chevrotain"

const LCurly = createToken({ name: "LCurly", pattern: /{/ })
const RCurly = createToken({ name: "RCurly", pattern: /}/ })
const LBrace = createToken({ name: "LBrace", pattern: /\(/ })
const RBrace = createToken({ name: "RBrace", pattern: /\)/ })
const Optional = createToken({ name: "Optional", pattern: /\?/ })
const Colon = createToken({ name: "Colon", pattern: /:/ })
const Comma = createToken({ name: "Comma", pattern: /,/ })
const Identifier = createToken({ name: "Identifier", pattern: /[a-zA-Z_][a-zA-Z0-9_]*/ })
const Number = createToken({ name: "Number", pattern: /\d+/ })
const LineBreak = createToken({ name: "LineBreak", pattern: /[\r\n]+/, group: Lexer.SKIPPED })
const WhiteSpace = createToken({ name: "WhiteSpace", pattern: /\s+/, group: Lexer.SKIPPED });

const allTokens = [ LineBreak, WhiteSpace, LCurly, RCurly, LBrace, RBrace, Optional, Colon, Comma, Identifier, Number ]


class SchemaParser extends CstParser {
  constructor() {
    super(allTokens)

    this.RULE("file", () => {
      this.MANY(() => this.SUBRULE((this as any).entry))
    })

    this.RULE("entry", () => {
      this.CONSUME(Identifier)
      this.CONSUME(LCurly)
      this.OPTION(() => this.MANY(() => this.SUBRULE((this as any).property)))
      this.CONSUME(RCurly)
    })

    this.RULE("property", () => {
      this.CONSUME(Identifier)
      this.OPTION(() => {
        this.CONSUME(LBrace)
        this.OPTION2(() => this.MANY(() => this.SUBRULE((this as any).attribute)))
        this.CONSUME(RBrace)
      })
      this.OPTION3(() => this.CONSUME(Optional))
      this.CONSUME(Colon)
      this.CONSUME2(Identifier)
      this.OPTION4(() => this.CONSUME(Comma))
    })

    this.RULE("attribute", () => {
      this.CONSUME(Identifier)
      this.OPTION(() => {
        this.CONSUME(Colon)
        this.SUBRULE((this as any).value)
      })
    })

    this.RULE("value", () => {
      this.OR([
        { ALT: () => this.CONSUME(Number) },
        { ALT: () => this.CONSUME(Identifier) },
      ])
    })

    this.performSelfAnalysis()
  }
}

const lexer = new Lexer(allTokens)
const parser = new SchemaParser()

interface Token {
  image: string,
  startColumn: number,
  startLine: number,
  startOffset: number,
  endColumn: number,
  endLine: number,
  endOffset: number,
  tokenType: {
    name: string
  },
}

interface Node<T> {
  name: string
  children: T
}

export type File = Node<{entry: Array<EntryNode>}>
type EntryNode = Node<{Identifier: [Token], property: Array<PropertyNode>}>
type PropertyNode = Node<{
  Identifier: [Token, Token],
  attribute?: Array<AttributeNode>,
  Optional?: Array<Token>}
>
type AttributeNode = Node<{Identifier: [Token], value?: [ValueNode]}>
type ValueNode = Node<{Number: [Token]} | {Identifier: [Token]}>


export function parse(input: string): File {
  const {errors, tokens} = lexer.tokenize(input)

  if (errors.length > 0) {
    throw errors
  }

  parser.input = tokens
  const result = (parser as any).file()

  if (parser.errors.length > 0) {
    throw parser.errors[0]
  }

  return result
}

export function formatError(input: string, error: any) {
  if (typeof Array.isArray(error)) {
    return error.map((x: { message: string, line: number, column: number }) =>
      (x.line > 2 ? input.split("\n")[x.line - 2] + "\n" : '') +
      input.split('\n')[x.line - 1] + "\n"
      + " ".repeat(x.column - 1) + "^ "
      + x.message
    ).join("\n")
  }
  if (error.token) {
    const { message, token: { startLine, startColumn } } = error as { message: string, token: { startLine: number, startColumn: number } }
    return [
      startLine > 2 && input.split("\n")[startLine - 2],
      input.split("\n")[startLine - 1],
      " ".repeat(startColumn - 1) + "^ ",
      `${message} at ${startLine},${startColumn}`,
    ].filter(x => x).join('\n')
  }
  return error
}

export function isNumber(value: ValueNode): value is Node<{Number: [Token]}> {
  return 'Number' in value.children
} 

export function isIdentifier(value: ValueNode): value is Node<{Identifier: [Token]}> {
  return 'Identifier' in value.children
} 
