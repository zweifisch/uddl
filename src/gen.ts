import {File, isIdentifier, isNumber} from './parser'

export interface Entry {
  name: string
  properties: Array<Property>
}

interface Property {
  key: string
  value: string
  attributes?: Array<Attribute>
}

interface Attribute {
  key: string
  value?: number | string
}

export function gen(input: File): Array<Entry> {
  return input.children.entry.map(entry => ({
    name: entry.children.Identifier[0].image,
    properties: entry.children.property.map(property => ({
      key: property.children.Identifier[0].image,
      value: property.children.Identifier[1].image,
      attributes: property.children.attribute?.map(attr => ({
        key: attr.children.Identifier[0].image,
        value: attr.children.value
          ? (isIdentifier(attr.children.value[0]) ? attr.children.value[0].children.Identifier[0].image :
            isNumber(attr.children.value[0]) ? parseFloat(attr.children.value[0].children.Number[0].image) : undefined)
          : undefined
      }))
    }))
  }))
}