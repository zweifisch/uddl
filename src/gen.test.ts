import {test, expect} from 'bun:test'
import { parse } from './parser'
import { gen } from './gen'

test('gen', () => {

  const input = `
User {
  id(primary_key auto_increment): int,
  name?: text,
  email(unique): text,
  gender(default:0): int,
}
`
  expect(gen(parse(input))).toMatchSnapshot()
})