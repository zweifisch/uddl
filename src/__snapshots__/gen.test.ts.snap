// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`gen 1`] = `
[
  {
    "name": "User",
    "properties": [
      {
        "attributes": [
          {
            "key": "primary_key",
            "value": undefined,
          },
          {
            "key": "auto_increment",
            "value": undefined,
          },
        ],
        "key": "id",
        "value": "int",
      },
      {
        "attributes": undefined,
        "key": "name",
        "value": "text",
      },
      {
        "attributes": [
          {
            "key": "unique",
            "value": undefined,
          },
        ],
        "key": "email",
        "value": "text",
      },
      {
        "attributes": [
          {
            "key": "default",
            "value": 0,
          },
        ],
        "key": "gender",
        "value": "int",
      },
    ],
  },
]
`;
